### PR TITLE
Specify authflow selection

### PR DESCRIPTION
**Pull request by IniZio**
ref #3921 / DEV-1083

### CODE REVIEW

File: docs/specs/authentication-flow.md [hunk 1]
```diff
@@ -356,6 +356,58 @@ reauth_flows:
...
+      # ...
+    - name: email_password_2fa
+      steps:
+      # ...
+
+oauth:
+  clients:
+    - client_id: public_app
+      # Only allow OAuth login for public_app
+      x_authentication_flows_allowlist:
```
Comment (IniZio) [inline docs/specs/authentication-flow.md:393]: Not sure if this need to apply on the `login_flow:` in account linking as well?

File: docs/specs/authentication-flow.md [hunk 2]
```diff
@@ -356,6 +356,58 @@ reauth_flows:
...
+      - type: login
+        name: oauth_only
+      - type: signup_login
+        name: oauth_only
+    - client_id: internal
+      # Allow both OAuth login and email_password_2fa for internal
+      x_authentication_flows_allowlist:
```
Comment (IniZio) [inline docs/specs/authentication-flow.md:400]: Not sure if one flow is less secure than the other, should we need to allow triggering a stricter flow on important actions e.g. delete account

File: docs/specs/sdk.md [hunk 3]
```diff
@@ -151,6 +151,62 @@ async function onClickSave() {
...
+
+The SDK allows the developer to choose custom authentication flow using the `flow` field in `AuthorizeOptions` and `ReauthenticateOptions`.
+
+It is mutually exclusive with the `page` field, which is used to specify whehter the login or signup page should be shown.
+
+```typescript
+interface AuthorizeOptions {
+  // ... fields omitted for brevity.
+
+  flow: {
```
Comment (IniZio) [inline docs/specs/sdk.md:164]: Will `flow` be confusing? Or maybe `customFlow` or `authenticationFlow`

File: docs/specs/authentication-flow.md (hunk 1 above)
Comment (IniZio) [reply docs/specs/authentication-flow.md]: Outdated.

File: docs/specs/authentication-flow.md [hunk 4]
```diff
@@ -356,6 +357,104 @@ reauth_flows:
...
+      steps:
+      # ...
+```
+
+Authentication flow groups need to be defined and whitelisted per client using the following configuration:
+
+```yaml
+ui:
+  authentication_flow:
+    groups:
```
Comment (IniZio) [inline docs/specs/authentication-flow.md:417]: Or should it be defined at `/authentication_flow/groups` instead?

File: docs/specs/authentication-flow.md [hunk 5]
```diff
@@ -356,6 +357,104 @@ reauth_flows:
...
+    - client_id: internal
+      # Allow both OAuth login and email_password_2fa for internal
+      x_authentication_flow_group_allowlist:
+      - oauth_only
+      - email_password_2fa
+```
+
+To authorize using a flow group, use `x_authentication_flow_group` and `x_page` in the [authentication request](/docs/specs/oidc.md#x_authentication_flow_group)
+
+For clients with `x_authentication_flow_group_allowlist` specified, only the specified flow groups can be used.
```
Comment (IniZio) [inline docs/specs/authentication-flow.md:446]: If not specified, should client only be able to use `default` or no constraint? Since currently default UI always use `default`

File: docs/specs/authentication-flow.md [hunk 6]
```diff
@@ -356,6 +357,104 @@ reauth_flows:
...
+      # ...
+```
+
+Authentication flow groups need to be defined and whitelisted per client using the following configuration:
+
+```yaml
+ui:
+  authentication_flow:
+    groups:
+    - name: oauth_only
```
Comment (IniZio) [inline docs/specs/authentication-flow.md:418]: Can `groups[].name = "default` be overridden or reserved?

File: docs/specs/authentication-flow.md (hunk 5 above)
Comment (louischan-oursky) [reply docs/specs/authentication-flow.md:446]: Should be no constraint, so that custom UI can use flows other than `default`.

File: docs/specs/authentication-flow.md (hunk 6 above)
Comment (louischan-oursky) [reply docs/specs/authentication-flow.md:418]: Can be overridden.

File: docs/specs/oidc.md [hunk 7]
```diff
@@ -221,12 +223,28 @@ Only `S256` is supported. `plain` is not supported.
 
 To mitigate replay attacks, provide a `nonce` in the authentication request. Authgear will include the `nonce` Claim in the ID Token, and the client must verify that the `nonce` claim value is equal to the value of the `nonce` parameter sent in the authentication request. The `nonce` is recommended but it is optional. The `nonce` value is a case sensitive string. Reference: [Authentication Request](https://openid.net/specs/openid-connect-core-1_0.html#rfc.section.3.1.2.1).
 
+### x_page
+
+It specifies either the login or signup page to be shown to the user.
+
+This option is mutually exclusive with `x_authentication_flow_type`. If both are specified, the request will be rejected.
```
Comment (louischan-oursky) [inline docs/specs/oidc.md:230]: This line is outdated.

File: docs/specs/sdk.md [hunk 8]
```diff
@@ -151,6 +151,71 @@ async function onClickSave() {
...
+It can be used in conjunction with `page` field in `AuthorizeOptions` to specify whether to use login or signup flow.
+
+```typescript
+interface AuthorizeOptions {
+  // ... fields omitted for brevity.
+
+  page?: "login" | "signup";
+  authflow?: {
+    group?: string;
+  };
```
Comment (louischan-oursky) [inline docs/specs/sdk.md:167]: authenticationFlowGroup?: string;

File: e2e/pkg/testrunner/runner.go [hunk 9]
```diff
@@ -33,6 +33,10 @@ func (tr *TestRunner) Run() {
 	hasFocus := false
 	for _, testCase := range testCases {
 		if testCase.Focus {
+			if os.Getenv("CI") != "" {
+				t.Fatal("focus test case in CI")
+			}
+
```
Comment (louischan-oursky) [inline e2e/pkg/testrunner/runner.go:39]: This change is unrelated to this PR.

File: docs/specs/oidc.md [hunk 10]
```diff
@@ -221,12 +223,28 @@ Only `S256` is supported. `plain` is not supported.
...
 
 Supported values: `change_password`.
 
+### x_authentication_flow_group
+
+When it is specified, the user perform authentication through custom authentication flow group.
+
+It can be used in conjunction with `x_page` to decide whether to execute login or signup flow.
+
+If the specified group is not found or not included in the [client's allow list](/docs/specs/authentication-flow.md#flow-selection), the request will be rejected.
```
Comment (louischan-oursky) [inline docs/specs/oidc.md:246]: Change the description to 

> Override the authentication flow group in the default UI. For details, see [insert a proper link here].

File: docs/specs/authentication-flow.md [hunk 11]
```diff
@@ -356,6 +357,104 @@ reauth_flows:
...
+
+oauth:
+  clients:
+    - client_id: public_app
+      # Only allow OAuth login for public_app
+      x_authentication_flow_group_allowlist:
+      - oauth_only
+    - client_id: internal
+      # Allow both OAuth login and email_password_2fa for internal
+      x_authentication_flow_group_allowlist:
```
Comment (louischan-oursky) [inline docs/specs/authentication-flow.md:440]: Mention this new client metadata in `docs/specs/oidc.md#custom-client-metadata`. The description is a backlink to a proper section in authentication-flow-in-default-ui.md

**Comment by IniZio** [comment]
Will also add constraint for custom UI and authflow API

File: docs/specs/authentication-flow-selection.md [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+```yaml
+ui:
+  authentication_flow:
+    groups:
+    - name: oauth_only
+      login_flow: oauth_only
+      signup_flow: oauth_only
+      reauth_flow: password
+      promote_flow: password
+      account_recovery_flow: email
```
Comment (louischan-oursky) [inline docs/specs/authentication-flow-selection.md:95]: should have signup_login_flow

File: docs/specs/authentication-flow-selection.md [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+{
+  // ... other fields of the authorization request
+  "x_page": "login",
+  "x_authentication_flow_group": "oauth_only"
+}
+```
+
+### Flow Allowlist
+
+Each client app may have an allowlist of flows defined with `x_authentication_flow_allowlist`. Once specified, the authorization request will be rejected if specified flow is not in the allowlist.
```
Comment (louischan-oursky) [inline docs/specs/authentication-flow-selection.md:130]: The authorization request will not rejected, as the flow name parameter is consumed by the authflow HTTP API. Instead, the authflow HTTP API will return error, when the authflow is created with a OAuth Session of a particular client ID.

Comment (IniZio) [reply docs/specs/authentication-flow-selection.md]: Updated to be specific to authflow api

**Comment by IniZio** [comment]
Updated spec to merge flow group and flow allowlist, and use list for each flow type in flow group 🙏

File: docs/specs/authentication-flow-selection.md [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+You can specify allowed flow groups and individual flows for each client app:
+
+```yaml
+oauth:
+  clients:
+  - client_id: public_app
+    x_authentication_flow_allowlist:
+      # Only allow OAuth login for public_app
+      groups:
+      - oauth_only
```
Comment (louischan-oursky) [inline docs/specs/authentication-flow-selection.md:107]: Rename to `- name: oauth_only`. That is, change from a plain string to a struct with a field called `name`.

**Review by louischan-oursky — APPROVED** [review]
